resources/aws/instance.go

```diff
@@ -42,60 +42,90 @@ type Instance struct {
 	AWSEntity
 }
 
-func existingInstanceID(instances ec2.DescribeInstancesOutput) (string, error) {
-	for _, r := range instances.Reservations {
-		for _, i := range r.Instances {
-			return *i.InstanceId, nil
-		}
+func statePendingOrRunning(instance *ec2.Instance) bool {
+	stateCode := *instance.State.Code
+	switch stateCode {
+	case int64(EC2PendingState), int64(EC2RunningState):
+		return true
 	}
-	return "", microerror.MaskAny(instanceFindError)
+
+	return false
 }
 
-func (i *Instance) CreateIfNotExists() (bool, error) {
-	instances, err := i.Clients.EC2.DescribeInstances(&ec2.DescribeInstancesInput{
-		Filters: []*ec2.Filter{
-			&ec2.Filter{
-				Name: aws.String(fmt.Sprintf("tag:%s", tagKeyName)),
-				Values: []*string{
-					aws.String(i.Name),
-				},
+func (i Instance) findExisting() (*ec2.Instance, error) {
+	filters := []*ec2.Filter{}
+	if i.ClusterName != "" {
+		filters = append(filters, &ec2.Filter{
+			Name: aws.String(fmt.Sprintf("tag:%s", tagKeyCluster)),
+			Values: []*string{
+				aws.String(i.ClusterName),
 			},
-			&ec2.Filter{
-				Name: aws.String(fmt.Sprintf("tag:%s", tagKeyCluster)),
-				Values: []*string{
-					aws.String(i.ClusterName),
-				},
-			},
-			&ec2.Filter{
-				Name: aws.String("instance-state-code"),
-				Values: []*string{
-					aws.String(fmt.Sprintf("%d", EC2RunningState)),
-				},
+		})
+	}
+	if i.Name != "" {
+		filters = append(filters, &ec2.Filter{
+			Name: aws.String(fmt.Sprintf("tag:%s", tagKeyName)),
+			Values: []*string{
+				aws.String(i.Name),
 			},
-			&ec2.Filter{
-				Name: aws.String("instance-state-code"),
-				Values: []*string{
-					aws.String(fmt.Sprintf("%d", EC2PendingState)),
-				},
+		})
+	}
+	if i.id != "" {
+		filters = append(filters, &ec2.Filter{
+			Name: aws.String("instance-id"),
+			Values: []*string{
+				aws.String(i.id),
 			},
-		},
+		})
+	}
+
+	reservations, err := i.Clients.EC2.DescribeInstances(&ec2.DescribeInstancesInput{
+		Filters: filters,
 	})
 	if err != nil {
-		return false, microerror.MaskAny(err)
+		return nil, microerror.MaskAny(err)
 	}
 
-	if len(instances.Reservations) > 0 {
-		instanceID, err := existingInstanceID(*instances)
-		if err != nil {
-			return false, microerror.MaskAny(err)
+	for _, reservation := range reservations.Reservations {
+		for _, instance := range reservation.Instances {
+			if statePendingOrRunning(instance) {
+				return instance, nil
+			}
+		}
+	}
+
+	return nil, microerror.MaskAny(instanceFindError)
+}
+
+func (i *Instance) checkIfExists() (bool, error) {
+	instance, err := i.findExisting()
+	if err != nil {
+		if strings.Contains(err.Error(), instanceFindError.Error()) {
```

Comment:
We agreed on not comparing strings. Please use an error matcher. See for instance https://github.com/giantswarm/kubernetesd/blob/master/service/error.go.